modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Assets.scala

```diff
@@ -0,0 +1,289 @@
+    files: Ref[F, Map[URI, Hash]],
+    stop: Ref[F, Boolean],
+    clients: Clients[F]
+  )
+
+  object State {
+
+    /** Test pair is used in tests to initialize HTTP client */
+    private val TestPair: Asset = URI.create("http://localhost:8080") -> "index"
+
+    /**
+     * Initialize an assets state. Try to find them on local FS
+     * or download if they're missing. Also initializes all necessary
+     * clients (S3, GCP, HTTP etc)
+     * @param blocker thread pool for downloading and reading files
+     * @param stop global stop signal from [[Environment]]
+     * @param assets all assets that have to be tracked
+     */
+    def make[F[_]: ConcurrentEffect: Timer: ContextShift](
+      blocker: Blocker,
+      stop: Ref[F, Boolean],
+      assets: List[Asset]
+    ): Resource[F, State[F]] =
+      for {
+        clients <- Clients.make[F](blocker, assets.map(_._1))
+        map <- Resource.liftF(build[F](blocker, clients, assets.filterNot(asset => asset == TestPair)))
+        files <- Resource.liftF(Ref.of[F, Map[URI, Hash]](map))
+      } yield State(files, stop, clients)
+
+    def build[F[_]: Concurrent: Timer: ContextShift](
+      blocker: Blocker,
+      clients: Clients[F],
+      assets: List[Asset]
+    ): F[Map[URI, Hash]] =
+      buildFromLocal(blocker, assets)
+        .flatMap { hashes =>
+          hashes.traverse {
+            case (uri, path, Some(hash)) =>
+              Logger[F].info(s"Asset from $uri is found on local system at $path").as(uri -> hash)
+            case (uri, path, None) =>
+              downloadAndHash[F](clients, blocker, uri, Paths.get(path)).map(hash => uri -> hash)
+          }
+        }
+        .map(_.toMap)
+
+    def buildFromLocal[F[_]: Sync: ContextShift](blocker: Blocker, assets: List[Asset]): F[List[(URI, String, Option[Hash])]] =
+      assets.traverse { case (uri, path) => local[F](blocker, path).map(hash => (uri, path, hash)) }
+
+    /** Check if file already exists */
+    def local[F[_]: Sync: ContextShift](blocker: Blocker, path: String): F[Option[Hash]] = {
+      val fpath = Paths.get(path)
+      exists(blocker, fpath).ifM(
+        readAll(fpath, blocker, 1024).through(md5).compile.to(Array).map(b => Hash(b).some),
+        Sync[F].pure(none)
+      )
+    }
+  }
+
+  final case class Hash private (s: String) extends AnyVal
+
+  object Hash {
+    def apply(bytes: Array[Byte]): Hash = {
+      val bi = new java.math.BigInteger(1, bytes)
+      Hash(String.format("%0" + (bytes.length << 1) + "x", bi))
+    }
+  }
+
+  /** Pair of a tracked `URI` and destination path on local FS (`java.nio.file.Path` is not serializable) */
+  type Asset = (URI, String)
+
+  /** Initialise the [[updateStream]] with all necessary resources if refresh period is specified */
+  def run[F[_]: ConcurrentEffect: ContextShift: Timer: Parallel](env: Environment[F]): Stream[F, Unit] =
+    env.assetsUpdatePeriod match {
+      case Some(duration) =>
+        val init = for {
```

Comment:
Don't we want to pause enrich here?

Reply:
It's paused by default, when the whole thing starts https://github.com/snowplow/enrich/blob/feature/fs2-enrich/modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Environment.scala#L125